server/routes/users.ts

```diff
@@ -16,46 +18,94 @@ router.get('/signOut', async (req, res) => {
+        idToken = idToken.split('JWT ')[1];  
+        const decodedToken: DecodedIdToken = await getAuth().verifyIdToken(idToken);
+        uid = decodedToken.uid;
+        email = decodedToken.email;
     } catch(e) {
         return res.status(403).json({ message: 'Unauthorized. Failed to verify token.', error: e.message });
     }
+
+    try { // ensure user is in the database
+        await getUserByEmail(email);
+    } catch(e) {
+        return res.status(401).json({ message: 'User does not exist in database, try sign up.', error: e.message });
+    }
+
+    // establish express session for current user
+    const user = { _id: uid, email: email };
+    req.session.user = user;
+    return res.json({ message: 'Signed in.' });
 });
 
 // POST /signUp
 router.post('/signUp', async (req, res) => {
-    if(req.session.user)
+    if(req.session.user) // already logged in
         return res.status(403).json({ message: 'Unauthorized. User already logged in.' });
 
-    if(!req.headers.authorization)
+    if(!req.headers.authorization) // no auth token provided
         return res.status(403).json({ message: 'Unauthorized. Need token for login.' });
 
-    try {
+    let uid: string, email: string, name: string;
+    try { // authenticate with firebase
         let idToken: string = req.headers.authorization as string;
         idToken = idToken.split('JWT ')[1];
-        const { email, uid } = await getAuth().verifyIdToken(idToken);
-        const { displayName } = await getAuth().getUser(uid);
-        // TODO: check (uid) to make sure it hasn't been used already in the DB.
-        // TODO: pass (uid, email, displayName) to a createUser function in data/users.ts to populate DB.
-        const user = { _id: uid, email: email };
-        req.session.user = user;
-        return res.json({ message: 'Signed up.' });
+        if(!idToken || !checkNonEmptyString(idToken))
+            throw new Error("[routes/users POST signUp] id token is invalid.");
+        const decodedToken: DecodedIdToken = await getAuth().verifyIdToken(idToken);
+        uid = decodedToken.uid;
+        email = decodedToken.email;
+        name = (await getAuth().getUser(uid)).displayName;
     } catch(e) {
         return res.status(403).json({ message: 'Unauthorized. Failed to verify token.', error: e.message });
     }
+
+    try { // ensure user does not exist already
+        await getUserByEmail(email);
+        return res.status(401).json({ message: 'User already exists. Cannot sign up with this email' });
+    } catch(e) {}
+
+    try { // Add the user to the DB
+        await createUser(uid, email, name);
+    } catch(e) {
+        return res.status(500).json({ message: 'Issue creating user in database.', error: e.message });
+    }
+
+    // establish express session for current user
+    const user = { _id: uid, email: email };
+    req.session.user = user;
+    return res.json({ message: 'Signed up.' });
```

Comment:
Could lines 85-88 be placed inside the try-catch starting on line 79?

Reply:
I believe try-catch blocks should be as small as possible to encapsulate only the essence of the code that could fail. This way, it is easier to see what is causing issues instead of putting everything in a single try-catch